config.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/knadh/koanf"
+	"github.com/knadh/koanf/parsers/yaml"
+	"github.com/knadh/koanf/providers/env"
+	"github.com/knadh/koanf/providers/file"
+	"github.com/knadh/koanf/providers/posflag"
+	flag "github.com/spf13/pflag"
+)
+
+type configuration struct {
+	RefreshInterval int                 `koanf:"interval"`
+	Secret          secretConfiguration `koanf:"secret"`
+	Oidc            oidcConfiguration   `koanf:"oidc"`
+	Log             logConfiguration    `koanf:"log"`
+
+	DummyProvider bool `koanf:"dummyprovider"`
+}
+
+type secretConfiguration struct {
+	Name      string `koanf:"name"`
+	Namespace string `koanf:"namespace"`
+	Key       string `koanf:"key"`
+}
+
+type logConfiguration struct {
+	Level  string `koanf:"level"`
+	Format string `koanf:"format"`
+}
+
+const (
+	TextFormat = "text"
+	JSONFormat = "json"
+
+	LevelDebug = "debug"
+	LevelInfo  = "info"
+	LevelWarn  = "warn"
+)
+
+type oidcConfiguration struct {
+	TokenUrl     string `koanf:"tokenurl"`
+	ClientID     string `koanf:"clientid"`
+	ClientSecret string `koanf:"clientsecret"`
+}
+
+func getConfig() (configuration, error) {
+	c := configuration{
+		Secret: secretConfiguration{
+			Key: "token",
+		},
+		RefreshInterval: 500,
+		DummyProvider:   false,
+		Log: logConfiguration{
+			Level:  LevelInfo,
+			Format: TextFormat,
+		},
+	}
+	k := koanf.New(".")
+
+	f := flag.NewFlagSet("config", flag.ContinueOnError)
+	configFile := f.StringP("config", "f", "", "path to configuration file")
+	f.String("secret.name", "", "name of the secret to update")
+	f.String("secret.namespace", "", "namespace of the secret to update")
+	f.String("secret.key", c.Secret.Key, "name of the secret to update")
+
+	f.Int("interval", c.RefreshInterval, "interval in seconds to update the token")
+	f.String("log.level", c.Log.Level, "the log level, on of [debug, info, warn]")
+	f.String("log.format", c.Log.Format, "how to format the log, on of [text, json]")
+
+	f.String("oidc.tokenurl", "", "the OIDC token endpoint")
+	f.String("oidc.clientid", "", "the OIDC client id")
+	f.String("oidc.clientsecret", "", "the OIDC client secret")
+	f.Parse(os.Args[1:])
+
+	if *configFile != "" {
+		if err := k.Load(file.Provider(*configFile), yaml.Parser()); err != nil {
+			return c, fmt.Errorf("unable to read file %s: %w", *configFile, err)
+		}
+	}
+
+	if err := k.Load(env.Provider("KTR_", ".", func(s string) string {
+		return strings.ReplaceAll(strings.ToLower(
+			strings.TrimPrefix(s, "KTR_")), "_", ".")
```

Comment:
Did you double-check whether the `_` is sufficient for a level separator?
e.g. does `KTR_OIDC_TOKENURL` actually work? what if there's a name like `KTR_OIDC_TOKEN_URL`, would `URL` be part of TOKEN_URL or be on another level? (let's suppose you have another level)

This issue is why I previously linked to a solution where the level separator is a double underscore: `__`, so that `KTR_OIDC__TOKEN_URL` would be clear to recognize `token-url`. I know it's not the case here but maybe in the future we avoid breaking changes.

Reply:
The current setup works. I will break if we add underscores in keys.

I personally don't like double underscores, as in my opinion they are hard to read and depending on font can be mistaken as a single underscore.

But if you really prefer the double underscore I can switch to it.

Reply:
To answer you question. I think `KTR_OIDC_TOKEN_URL` would resolve to the key `oidc.token.url` so the `url` would be on another level. If we introduce a `oidc.token_url` key we would not be able to set it using environment variables. 

I don't really see a good solution. As I said I'm not a fan of double underscores, but can be convinced otherwise.

Reply:
it's fine by me. let's leave it for now :)